Doc: Incorporate field ref deep dive content (#17484)
elastic/logstash: docs/reference/event-dependent-configuration.md

You might find situations in which you need to refer to a field or collection of fields by name. 
You can accomplish this goal using the Logstash field reference syntax.

The syntax to access a field specifies the entire path to the field, with each fragment wrapped in square brackets.
When a field name contains square brackets, the brackets must be properly [ escaped](#formal-grammar-escape-sequences).

Field references can be expressed literally within [conditional statements](#conditionals) in your pipeline configurations,
as string arguments to your pipeline plugins, or within sprintf statements that will be used by your pipeline plugins:

```
filter {
  #  +----literal----+     +----literal----+
  #  |               |     |               |
  if [@metadata][date] and [@metadata][time] {
    mutate {
      add_field {
        "[@metadata][timestamp]" => "%{[@metadata][date]} %{[@metadata][time]}"
      # |                      |    |  |               |    |               | |
      # +----string-argument---+    |  +--field-ref----+    +--field-ref----+ |
      #                             +-------- sprintf format string ----------+
      }
    }
  }
}
```

> ### Formal grammar [formal-grammar]

Reviewer: Maybe this section (from here until `Escape Sequences`) could be included as a drop-down, and/or at the bottom of the field reference section, with the `Escape Sequences` section lifted up.  

This is a highly technical discussion and, while informative, may not be best suited to be in the main flow of ways to access event data and fields in a pipeline

> Expand the section below if you want more details about formal grammar for field references.

Reviewer: I love the drop-down, I think it works well, and ensures that the standard user flow is uninterrupted

I'm not quite sure what the right answer is here, but I'm not sure if we need this line, or if it could be re-worded a little to reduce the repetition of "Formal Grammar for field references."

```suggestion
Expand the section below if you would like to explore the formal grammar.
```

A _Field Reference Literal_ is a sequence of one or more _Path Fragments_ that can be used directly in Logstash pipeline [conditionals](#conditionals)  without any additional quoting.
Example: `[request]`, `[response][status]`).

```
fieldReferenceLiteral
  : ( pathFragment )+
  ;
```

> NOTE: In Logstash 7.x and earlier, a quoted value (such as `["foo"]`) is considered a field reference and isn't treated as a single element array.
> This behavior might cause confusion in conditionals, such as `[message] in ["foo",
> "bar"]` compared to `[message] in ["foo"]`.
> We discourage using names with quotes, such as `"\"foo\""`, as this behavior might change in the future.

Author: ```suggestion
```

Author: As of the 9.0 publishing, this seems like it can/should come out.
